cmd/firectl/main.go

```diff
@@ -105,6 +109,74 @@ func parseVsocks(devices []string) ([]firecracker.VsockDevice, error) {
 	return result, nil
 }
 
+func createFifoFileLogs(fifoPath string) (*os.File, error) {
+	return os.OpenFile(fifoPath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
+}
+
+func handleFifos(opts options) (io.Writer, []func() error, error) {
+	// these booleans are used to check whether or not the fifo queue or metrics
+	// fifo queue needs to be generated. If any which need to be generated, then
+	// we know we need to create a temporary directory. Otherwise, a temporary
+	// directory does not need to be created.
+	generateFifoFilename := false
+	generateMetricFifoFilename := false
+	cleanupFns := []func() error{}
+	var err error
+
+	var fifo io.WriteCloser
+	if len(opts.FcFifoLogFile) != 0 {
+		// if a fifo log file was specified via the CLI then we need to check if
+		// the log or metric fifo was also specified. If not, we will then generate
+		// those fifo files.
+		if len(opts.FcLogFifo) == 0 {
+			generateFifoFilename = true
+		}
+
+		if len(opts.FcMetricsFifo) == 0 {
+			generateMetricFifoFilename = true
+		}
+
+		if fifo, err = createFifoFileLogs(opts.FcFifoLogFile); err != nil {
+			return fifo, cleanupFns, fmt.Errorf("Failed to create fifo log file: %v", err)
+		}
+
+		cleanupFns = append(cleanupFns, func() error {
+			return fifo.Close()
+		})
+	} else if len(opts.FcLogFifo) > 0 || len(opts.FcMetricsFifo) > 0 {
+		// this checks to see if either one of the fifos was set. If at least one
+		// has been set we check to see if any of the others were not set. If one
+		// isn't set, we will generate the proper file path.
+		if len(opts.FcLogFifo) == 0 {
+			generateFifoFilename = true
+		}
+
+		if len(opts.FcMetricsFifo) == 0 {
+			generateMetricFifoFilename = true
+		}
+	}
+
+	if generateFifoFilename || generateMetricFifoFilename {
+		dir, err := ioutil.TempDir(os.TempDir(), "fcfifo")
+		if err != nil {
+			return fifo, cleanupFns, fmt.Errorf("Fail to create temporary directory: %v", err)
+		}
+
+		cleanupFns = append(cleanupFns, func() error {
+			return os.RemoveAll(dir)
+		})
+		if generateFifoFilename {
+			opts.FcLogFifo = filepath.Join(dir, "fc_fifo")
+		}
+
+		if generateMetricFifoFilename {
+			opts.FcMetricsFifo = filepath.Join(dir, "fc_metrics_fifo")
+		}
+	}
+
+	return fifo, cleanupFns, nil
```

Comment:
I think this can be simplified:
```go
func handleFifos(opts options) (io.Writer, []func() error, error) {
	var (
		generateFifoFilename       = len(opts.FcLogFifo) == 0
		generateMetricFifoFilename = len(opts.FcMetricsFifo) == 0
		cleanupFns                 = []func() error{}
		err                        error
		fifo                       io.WriteCloser
	)

	if len(opts.FcFifoLogFile) > 0 {
		if fifo, err = createFifoFileLogs(opts.FcFifoLogFile); err != nil {
			return fifo, cleanupFns, fmt.Errorf("Failed to create fifo log file: %v", err)
		}

		cleanupFns = append(cleanupFns, func() error {
			return fifo.Close()
		})
	}

	if generateFifoFilename || generateMetricFifoFilename {
		// ...
	}

	return fifo, cleanupFns, nil
}

```

Reply:
According to your logic, if `opts.FcLogFifo` not empty, `generateFifoFilename` will be true, no matter which branch it'll go, you do same checks in both branches. Do I missing something?

Reply:
I think I get it :)

```go
if len(opts.FcLogFifo) > 0 || len(opts.FcMetricsFifo) > 0 || len(opts.FcFifoLogFile) > 0 {
                generateFifoFilename = len(opts.FcLogFifo) == 0
		generateMetricFifoFilename = len(opts.FcMetricsFifo) == 0
}

if len(opts.FcFifoLogFile) > 0 {
               createFifoFileLogs()
               ...
}

```

Reply:
I am going to leave the function as is since I think it is easier to read.